polkadot/node/core/prospective-parachains/src/lib.rs

```diff
@@ -323,13 +334,27 @@ fn prune_view_candidate_storage(view: &mut View, metrics: &Metrics) {
 	let active_leaves = &view.active_leaves;
 	let mut live_candidates = HashSet::new();
 	let mut live_paras = HashSet::new();
-	for sub_view in active_leaves.values() {
-		for (para_id, fragment_tree) in &sub_view.fragment_trees {
-			live_candidates.extend(fragment_tree.candidates());
+	for (leaf, sub_view) in active_leaves.iter() {
+		live_candidates.extend(sub_view.pending_availability.iter().cloned());
+
+		for (para_id, fragment_chain) in &sub_view.fragment_chains {
+			live_candidates.extend(fragment_chain.candidates());
 			live_paras.insert(*para_id);
-		}
 
-		live_candidates.extend(sub_view.pending_availability.iter().cloned());
+			if let Some(storage) = view.candidate_storage.get(para_id) {
+				let unconnected_potential =
+					fragment_chain.find_unconnected_potential_candidates(storage, None);
+				if !unconnected_potential.is_empty() {
+					gum::trace!(
+						target: LOG_TARGET,
+						?leaf,
+						"Keeping unconnected candidates in storage: {:?}",
+						unconnected_potential
+					);
+				}
+				live_candidates.extend(unconnected_potential);
+			}
+		}
 	}
 
 	view.candidate_storage.retain(|para_id, storage| {
```

Comment:
The garbage collection, is really non-obvious here, could we add some debug logs  or metrics with the size of this objects. Thank you!

Reply:
done